contracts/timelocks/YakTimelockForDexStrategyV4.sol

```diff
@@ -0,0 +1,444 @@
+
+    constructor() {
+        manager = msg.sender;
+        feeCollector = msg.sender;
+    }
+
+    // Modifiers
+
+     /**
+     * @notice Restrict to `emergencyWithdraw`
+     * @dev To change, call revokeRole with emergencyWithdraw_ADMIN
+     */
+    modifier hasPermission (PermissionedTimelockFunctions _fn, address account) {
+        require(_hasAccess(_fn, account), "Caller is not emergencyWithdraw_ADMIN");
+        _;
+    }
+
+    /**
+     * @notice Restrict to `manager`
+     * @dev To change manager, deploy new timelock and transfer strategy ownership
+     */
+    modifier onlyManager {
+        require(msg.sender == manager);
+        _;
+    }
+
+    /**
+     * @notice Restrict to `feeCollector`
+     * @dev To change feeCollector, deploy new timelock and transfer strategy ownership
+     */
+    modifier onlyFeeCollector {
+        require(msg.sender == feeCollector);
+        _;
+    }
+
+    /**
+     * @notice Set timelock when changing pending values
+     * @param _strategy address
+     * @param _fn Function enum value
+     * @param timelockLength in seconds
+     */
+    modifier setTimelock(address _strategy, Functions _fn, uint timelockLength) {
+        timelock[_strategy][_fn] = block.timestamp + timelockLength;
+        _;
+    }
+
+    /**
+     * @notice Enforce timelock for a given function
+     * @dev Ends execution by resetting timelock to avoid replay
+     * @param _strategy address
+     * @param _fn Function enum value
+     */
+    modifier enforceTimelock(address _strategy, Functions _fn) {
+        require(timelock[_strategy][_fn] != 0 && timelock[_strategy][_fn] <= block.timestamp, "YakTimelockManager::enforceTimelock");
+        _;
+        timelock[_strategy][_fn] = 0;
+    }
+
+    function _hasAccess(PermissionedTimelockFunctions _ptf, address account) internal returns (bool) {
+        return authorizedAddressess[_ptf][account];
+    }
+
+    function grantAccountAccess(PermissionedTimelockFunctions _ptf, address account) external onlyManager {
+        require(!authorizedAddressess[_ptf][account], "Account has already been given access");
+        emit grant(_ptf,account);
+        authorizedAddressess[_ptf][account] = true;
+    }
+
+    function revokeAccountAccess(PermissionedTimelockFunctions _ptf, address account) external onlyManager {
+        require(!authorizedAddressess[_ptf][account], "Account has no access");
+        emit revoke(_ptf,account);
+        authorizedAddressess[_ptf][account] = false;
+    }
+
+    /**
+     * @notice Sweep tokens from the timelock to `feeCollector`
+     * @dev The timelock contract may receive assets from both revenue and asset recovery.
+     * @dev The sweep function is NOT timelocked, because recovered assets must go through separate timelock functions.
+     * @param tokenAddress address
+     * @param tokenAmount amount
+     */
+    function sweepTokens(address tokenAddress, uint tokenAmount) external onlyFeeCollector {
```

Comment:
we should move from feeCollector to hasPermission, and fee collector is just yet, another guy who has permission to call this function, do you oppose that @snow-farmer ?